app/apollo/resolvers/group.js

```diff
@@ -310,17 +327,43 @@ const groupResolvers = {
       const user = whoIs(me);
 
       try {
-        logger.info({ req_id, user, org_id, groupUuids, clusterIds }, `${queryName} validating`);
-
-        await validAuth(me, org_id, ACTIONS.MANAGE, TYPES.GROUP, queryName, context);
+        logger.info({req_id, user, org_id, groupUuids, clusterIds}, `${queryName} validating`);
 
         validateString( 'org_id', org_id );
         groupUuids.forEach( value => { validateString( 'groupUuids', value ); } );
         clusterIds.forEach( value => validateString( 'clusterIds', value ) );
 
-        const groups = await models.Group.find({org_id, uuid: {$in: groupUuids}});
-        if (groups.length < 1) {
-          throw new NotFoundError(context.req.t('None of the passed group uuids were found'));
+        const groups = await getAllowedResources(me, org_id, ACTIONS.MANAGE, TYPES.GROUP, queryName, context, null, groupUuids);
+        logger.info({req_id, user, org_id, groupUuids, clusterIds}, `${queryName} retrieved allowed resources`);
+
+        if (groups.length != groupUuids.length) {
+          throw new NotFoundError(context.req.t('One or more of the passed group uuids were not found'));
+        }
+
+        if (groups.length < 1) { throw new NotFoundError(context.req.t('None of the passed group uuids were found')); }
+
+        let allAllowedClusters = false;
+        try {
+          await validAuth(me, org_id, ACTIONS.READ, TYPES.CLUSTER, queryName, context);
+          allAllowedClusters = true;
+        }
+        catch(e){ // If exception thrown, user does NOT have auth to all resources of this type, and code must later filter based on fine grained auth
+        }
+        logger.info({req_id, user, org_id, groupUuids, clusterIds, allAllowedClusters}, `${queryName} validating - allAllowed: ${allAllowedClusters}`);
+
+        let foundClusters = await commonClusterSearch(models, {org_id}, { limit: 0, skip: 0, startingAfter: null });
+        foundClusters = foundClusters.filter(cluster => clusterIds.includes(cluster.cluster_id));
+
+        if (!allAllowedClusters){
+          foundClusters = await filterResourcesToAllowed(me, org_id, ACTIONS.READ, TYPES.CLUSTER, foundClusters, context);
+          logger.info({req_id, user, org_id}, `${queryName} filtered resources to allowed`);
+        }
+
+        if(!clusterIds.length){
+          throw new ValidationError(context.req.t('No cluster uuids were passed', {'clusterIds':clusterIds}), context);
+        }
```

Comment:
If you are going to add a check for whether the input `clusters` is not empty:
- Use `if( clusterIds.length < 1 )` rather than `if(!clusterIds.length)` -- more readable at the very least, we're trying to see if the length is zero, not trying to see if the length is 'false-y'
- This is a 'simple' check -- there's no reason to wait and check this only after the expensive database and authorization checks.  If included, it should be after `validateString` statements and before `clustersIds.forEach` validation around line 334.

Reply:
fixed

Reply:
Resolved